### PR TITLE
[azure-spring-data-cosmos] Turn java.math.Big{Decimal,Integer} into simple types

### PR DESCRIPTION
**Pull request by Xiphoseer**
# Description

This PR adds `java.math.BigDecimal` and `java.math.BigInteger` to the set of [CosmosSimpleTypes], thus extending #31417 to those types. A "simple type" in the context of `spring-data-commons` is one that will "not [be] recursively analyzed" (c.f. [AbstractMappingContext#setSimpleTypeHolder]).

Using one of these types in a POJO in Java 17 currently results in a `InaccessibleObjectException` as described in #36831. While the workaround mentioned there to use `--add-opens` works, I believe that is the wrong approach. Adding that flag will prevent spring-data-commons from causing the error by using reflection to find properties, but that doesn't mean that Jackson / the ObjectMapper will treat it as a nested POJO.

Quite the opposite actually: Jackson has a special DecimalNode for BigDecimal and several configuration options on how it is serialized to and serialized from JSON. Thus, it's very much treated as an _atomic_ type by Jackson, and by extension this library (azure-spring-data-cosmos) and should be a "simple type" in the context of the MappingContext. My understanding is that being a simple type does _not_ prevent converters being registered for it, so this is orthogonal to #38691 and any kind of default conversion to JSON primitives for the type.

As for the question of whether this is a bug in `spring-data-commons`: I do believe that Spring Data Commons should default to a behavior that does not try to use reflection on `java.*` classes that do not have public fields. On the other hand, it is an abstraction layer over multiple different implementations and only those can reasonably opt-in to a type being both simple and supported based on the underlying technology. Not all Spring Data Projects use a JSON / NoSQL datamodel.

The default of the `SimpleTypeHandler` API is a set, which elements can't be excluded from without re-doing the whole set. So my conclusion would be that it's possible to commit to `BigDecimal` and `BigInteger` being simple for azure-spring-data-cosmos today, while it's not nearly as clear for upstream Spring Data Commons. Removing them from `CosmosSimpleTypes` after they've been added upstream is also not a breaking change.

The / one relevant bug on `spring-data-commons` is https://github.com/spring-projects/spring-data-commons/issues/2844, but with neither accepted, it's been causing issues for us in the services we work with. In our case, the source data model we want to import into CosmosDB is based on an XML Schema using `xs:decimal`, so the natural java type to use as per JAXB is `java.math.BigDecimal`.

[CosmosSimpleTypes]: https://github.com/Xiphoseer/azure-sdk-for-java/blob/91e0c45dd809232afa20457796e43aab88eaa116/sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/mapping/CosmosSimpleTypes.java
[AbstractMappingContext#setSimpleTypeHolder]: https://javadoc.io/doc/org.springframework.data/spring-data-commons/latest/spring.data.commons/org/springframework/data/mapping/context/AbstractMappingContext.html#setSimpleTypeHolder(org.springframework.data.mapping.model.SimpleTypeHolder)

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by trande4884** [comment]
@saragluna can you please review?

**Comment by trande4884** [comment]
@Xiphoseer can you also add some tests that actually write/read to a database to ensure we are getting out what we put in? Thanks!

**Comment by Xiphoseer** [comment]
@microsoft-github-policy-service agree company="PRODYNA"

**Review by saragluna — COMMENTED** [review]
The change looks good to me, but we should run the IT to verify it will not bring any breaking change

**Comment by Xiphoseer** [comment]
> @Xiphoseer can you also add some tests that actually write/read to a database to ensure we are getting out what we put in? Thanks!

I tried this and would appreciate some input on it. First, as far as CONTRIBUTING goes, I tried to run IT tests against the emulator but hit rate limiting even when explicitly starting it with the flag to disable that. Secondly, the Azure Portal wizard for "Azure Cosmos DB" now has a bunch more options to pick from (Core vs MongoDB, Serverless, Rate Limiting) which can default to values incompatible with the IT suite, 1000 RU limit is one example. It would be helpful to update CONTRIBUTING for that.

I guess the tricky part is to get this right without introducing breaking changes. My current understanding is that with or without the SimpleType changes, Jackson will special case BigDecimal and store it as a number, but I'd like to double-check that again. This means that https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/modeling-data#numbers-in-json applies which talks about the fact that number will be "stored as JSON". In practice that seems to mean that JSON numbers are stored as IEEE 754 binary64 in the backend, not only some clients, so there is some loss of precision involved.

In particular, BigDecimal keeps around trailing zeroes after the decimal point (e.g. 10.50) via its scale but the roundtrip through cosmos turns that into the _mathematically_ equivalent 10.5, which doesn't compare as `equals` in the context of BigDecimal.

The solution to that would be to serialize the numbers as string, which is what spring-data-mongodb does via a converter, can also be achived locally using `@JsonFormat(shape = Shape.STRING)` or could be done in the object mapper via feature SERIALIZE_NUMBERS_AS_STRING (probably way too broad) or an AnnotationIntrospector (bit of a misnomer as per Jackson docs) for all BigDecimal types.

What I'm unsure about is whether that last point is
- orthogonal to this PR because it's not using field reflection either way
- possible to configure in the object mapper without breaking any users on java pre-16/17 or having added `--add-opens` as per the linked issues.

**Comment by Xiphoseer** [comment]
> I've triggered the IT pipeline.

Thanks!

> This is also my concern, since if a user already store their BigDecimal in the Cosmos DB using another version client, can the number be read using this version.

After the testing mentioned above, I'm fairly confident that the SimpleType changes will indeed not affect the actual serialization / deserialization.

**Comment by kushagraThapar** [comment]
Thanks @saragluna for the PR review, appreciate your help!
@trande4884 can you please run the spring integration tests on this pipeline, thanks!

**Comment by trande4884** [comment]
@Xiphoseer since you are having issues with IT tests if you give me access I can push some tests to this PR

**Comment by Xiphoseer** [comment]
> @Xiphoseer since you are having issues with IT tests if you give me access I can push some tests to this PR

The "allow edits by maintainers" is on for this PR, but presumably that's not sufficient for some reason and you need write access to the fork?

**Comment by trande4884** [comment]
@Xiphoseer apologies that I have not followed up on this sooner. There are now merge conflicts, if you could please resolve I will re-run the IT tests.

**Comment by Xiphoseer** [comment]
> @Xiphoseer apologies that I have not followed up on this sooner. There are now merge conflicts, if you could please resolve I will re-run the IT tests.

Just as a heads up: I'll probably not find time for this before next week.

**Review by trande4884 — APPROVED** [review]
LGTM

File: sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/domain/BigJavaMathTypes.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package com.azure.spring.data.cosmos.domain;
```
Comment (kushagraThapar) [inline sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/domain/BigJavaMathTypes.java:1]: missing the new file license header.

**Review by kushagraThapar — APPROVED** [review]
LGTM, except one minor comment, please add the new license header for one of the new files.

File: sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/domain/BigJavaMathTypes.java (hunk 1 above)
Comment (trande4884) [reply sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/domain/BigJavaMathTypes.java:3]: Done.

**Pull request merged**